lib/importer/backend.js

```diff
@@ -0,0 +1,300 @@
+const xlsx = require("node-xlsx").default;
+const crypto = require("crypto");
+const assert = require('node:assert').strict;
+
+// An implementation of the interface described in https://struct.register-dynamics.co.uk/trac/wiki/DataImporter/API
+
+// FIXME: State storage all in-memory for now...
+
+let sessionStore = new Map();
+let jobStore = new Map();
+
+function makeAccessToken(kind) {
+    return kind + "-" + crypto.randomBytes(16).toString("hex");
+}
+
+///
+/// Control Interface
+///
+
+// Create a new import session; returns session ID (sid)
+exports.CreateSession = () => {
+    let sid = makeAccessToken("s");
+    sessionStore[sid] = new Map();
+    return sid;
+};
+
+///
+/// Session Interface
+///
+
+// Set the filename of the file for an import session.
+exports.SessionSetFile = (sid, filename) => {
+    const data = xlsx.parse(filename);
+    let session = {
+        filename: filename,
+        data: data,
+        jobs: new Set(),
+        sheets: new Map(),
+    }
+    data.forEach((sheet) => {
+        session.sheets[sheet.name] = sheet.data;
+    });
+    sessionStore[sid] = session;
+};
+
+function max(x,y) {
+    if(x > y) return x;
+    else return y;
+}
+
+// Returns the input file structure, as an object with a property "sheets"
+// mapping sheet names to objects with "rows" and "columns" properties.
+exports.SessionGetInputStructure = (sid) => {
+    assert(sessionStore[sid]);
+    let data = sessionStore[sid].data;
+    let sheetStructure = new Map();
+    data.forEach((sheet) => {
+        sheetStructure[sheet.name] = {
+            rows: sheet.data.length,
+            columns: sheet.data.reduce(
```

Comment:
Might these be rowLength and columnLength?  When I first saw it I think it might be the actual row/column data.

Reply:
Hmmm, good point. How about I name the surrounding thing "sheetDimensions" and the function "SessionGetInputDimensions" to make it clear these are dimensions?